File: Day42/Part1/arena_echoes.py
from typing import List, Tuple, Dict
import heapq

# Practical Dijkstra over (node, attr) with pruning.
# Returns minimal total cost to reach node n-1 with attr=0 or -1 if not found.

def popcount(x: int) -> int:
    return x.bit_count()


def min_cost_to_zero_attr(n: int, edges: List[Tuple[int, int, int]]) -> int:
    adj: List[List[Tuple[int, int]]] = [[] for _ in range(n)]
    for u, v, a in edges:
        adj[u].append((v, a))

    # dist[node]: dict attr->best_cost
    dist: List[Dict[int, int]] = [dict() for _ in range(n)]
    best_node = [10**30] * n  # best cost reaching node with any attr

    CAP = 512
    MARGIN = 64 * 60  # allow states within this margin of the node's current best

    pq: List[Tuple[int, int, int]] = []  # (cost, node, attr)
    heapq.heappush(pq, (0, 0, 0))
    dist[0][0] = 0
    best_node[0] = 0

    while pq:
        cost, u, a = heapq.heappop(pq)
        if dist[u].get(a, 10**30) != cost:
            continue
        if u == n - 1 and a == 0:
            return cost
        for v, w in adj[u]:
            na = a ^ w
            ncost = cost + popcount(na)
            # pruning by node best
            if ncost > best_node[v] + MARGIN:
                continue
            prev = dist[v].get(na)
            if prev is None or ncost < prev:
                dist[v][na] = ncost
                if ncost < best_node[v]:
                    best_node[v] = ncost
                # cap dictionary size with soft pruning: if too big, drop worst entries
                if len(dist[v]) > CAP:
                    # remove entries with cost > best_node[v] + MARGIN
                    to_del = [key for key, val in dist[v].items() if val > best_node[v] + MARGIN]
                    for key in to_del:
                        if len(dist[v]) <= CAP:
                            break
                        del dist[v][key]
                heapq.heappush(pq, (ncost, v, na))
    return -1


if __name__ == "__main__":
    n = 3
    edges = [(0, 1, 3), (1, 2, 1), (0, 2, 2)]
    print(min_cost_to_zero_attr(n, edges))  # -1 for example

File: Day42/Part1/arena_echoes.cpp
#include <bits/stdc++.h>
using namespace std;

static inline int popcountll(unsigned long long x){
#if defined(_MSC_VER)
    return (int)__popcnt64(x);
#else
    return __builtin_popcountll(x);
#endif
}

long long min_cost_to_zero_attr(int n, const vector<tuple<int,int,unsigned long long>>& edges){
    vector<vector<pair<int,unsigned long long>>> adj(n);
    for(auto &e: edges){ int u,v; unsigned long long a; tie(u,v,a)=e; adj[u].push_back({v,a}); }

    const long long INF = (long long)9e18;
    vector<unordered_map<unsigned long long, long long>> dist(n);
    vector<long long> best_node(n, INF);
    const size_t CAP = 512;
    const long long MARGIN = 64LL * 60LL;

    using State = tuple<long long,int,unsigned long long>; // cost,node,attr
    priority_queue<State, vector<State>, greater<State>> pq;
    dist[0][0]=0; best_node[0]=0; pq.emplace(0,0,0ULL);

    while(!pq.empty()){
        auto [cost,u,a]=pq.top(); pq.pop();
        auto it = dist[u].find(a);
        if(it==dist[u].end() || it->second!=cost) continue;
        if(u==n-1 && a==0ULL) return cost;
        for(auto &p: adj[u]){
            int v=p.first; unsigned long long w=p.second;
            unsigned long long na = a ^ w;
            long long ncost = cost + popcountll(na);
            if(ncost > best_node[v] + MARGIN) continue;
            auto it2 = dist[v].find(na);
            if(it2==dist[v].end() || ncost < it2->second){
                dist[v][na] = ncost;
                if(ncost < best_node[v]) best_node[v] = ncost;
                if(dist[v].size() > CAP){
                    // soft prune overly bad states
                    vector<unsigned long long> del;
                    del.reserve(dist[v].size());
                    for(auto &kv: dist[v]) if(kv.second > best_node[v] + MARGIN) del.push_back(kv.first);
                    for(auto key: del){ if(dist[v].size() <= CAP) break; dist[v].erase(key); }
                }
                pq.emplace(ncost, v, na);
            }
        }
    }
    return -1;
}

int main(){
    int n=3; vector<tuple<int,int,unsigned long long>> edges={{0,1,3ULL},{1,2,1ULL},{0,2,2ULL}};
    cout << min_cost_to_zero_attr(n, edges) << "\n"; // -1
    return 0;
}
